scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala

```diff
@@ -0,0 +1,47 @@
+package com.twitter.scalding.beam_backend
+
+import com.stripe.dagon.Rule
+import com.twitter.scalding.Execution.{ToWrite, Writer}
+import com.twitter.scalding.typed._
+import com.twitter.scalding.{CFuture, Config, Execution, ExecutionCounters}
+import org.apache.beam.sdk.Pipeline
+import scala.concurrent.{ExecutionContext, Future}
+
+class BeamWriter(val beamMode: BeamMode) extends Writer {
+  private def pipeline: Pipeline = beamMode.pipeline
+  override def start(): Unit = ()
+  override def finished(): Unit = ()
+
+  def getForced[T](conf: Config, initial: TypedPipe[T])(implicit cec: ExecutionContext): Future[TypedPipe[T]] = ???
+
+  def getIterable[T](conf: Config, initial: TypedPipe[T])(implicit cec: ExecutionContext): Future[Iterable[T]] = ???
+
+  override def execute(
+    conf: Config,
+    writes: List[ToWrite[_]])(implicit cec: ExecutionContext): CFuture[(Long, ExecutionCounters)] = {
+    import Execution.ToWrite._
+    val planner = BeamPlanner.plan(conf, beamMode.sources)
+    val phases: Seq[Rule[TypedPipe]] = OptimizationRules.standardMapReduceRules
+    val optimizedWrites = ToWrite.optimizeWriteBatch(writes, phases)
+
+    def rec(l: List[OptimizedWrite[TypedPipe, _]]): Unit = {
+      if (l.nonEmpty) {
+        l.head match {
+          case OptimizedWrite(pipe, ToWrite.SimpleWrite(opt, sink)) => {
+            val pcoll = planner(opt).run(pipeline)
+            beamMode.sink(sink) match {
+              case Some(ssink) =>
+                ssink.write(pipeline, conf, pcoll)
+              case _ => throw new Exception(s"unknown sink: $sink when writing $pipe")
+            }
+            rec(l.tail)
+          }
+          case _ => rec(l.tail)
+        }
+      }
+    }
+    rec(optimizedWrites)
+    pipeline.run().waitUntilFinish()
```

Comment:
You don't want to block `execute`, instead you want to create `Promise`, and resolve it when corresponding pipeline finishes